src/mem_extable.S

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Very unoptimized memcpy() to/from guest memory functions, using the HLV/HSV instructions.
+
+// Adds the instruction at 'lbl' to the exception table.
+.macro add_extable lbl
+.pushsection .extable, "a"
+.balign      8
+.quad        \lbl
+.popsection
+.endm
+.option push
+.option arch, +h
+.section .text
+
+// memcpy() to a guest physical address using HSV.
+.global _copy_to_guest
+_copy_to_guest:
+    // handle_trap assumes t0 holds the address of where we want to jump to when we encounter
+    // a fault and will stick SCAUSE in t1.
+    la    t0, _ret_from_copy
+    // _ret_from_copy assumes the return value is in t2.
+    mv    t2, zero
+1:
+    beq   t2, a2, _ret_from_copy
+    lb    t3, (a1)
+2:
+    hsv.b t3, (a0)
+    add_extable 2b
+    addi  a0, a0, 1
+    addi  a1, a1, 1
+    addi  t2, t2, 1
+    j     1b
+
+// memcpy() from a guest physical address using HLV.
+.global _copy_from_guest
+_copy_from_guest:
+    // handle_trap assumes t0 holds the address of where we want to jump to when we encounter
+    // a fault and will stick SCAUSE in t1.
+    la    t0, _ret_from_copy
+    // _ret_from_copy assumes the return value is in t2.
+    mv    t2, zero
+1:
+    beq   t2, a2, _ret_from_copy
+2:
+    hlv.b t3, (a1)
```

Comment:
```rust
_copy_from_guest(buf.as_mut_ptr(), gpa as *const u8, buf.len())
```
When reading GPA from guest, make sure guest's `vsatp` is set to mode 0 (Bare), othervise we are reading GVA instead of GPA.